pkg/chains/storage/grafeas/grafeas_test.go

```diff
@@ -0,0 +1,350 @@
+				tr:     test.args.tr,
+				client: client,
+				cfg: config.Config{
+					Storage: config.StorageConfigs{
+						Grafeas: config.GrafeasConfig{
+							ProjectID: "test-project",
+							NoteID:    "test-note",
+						},
+					},
+				},
+			}
+
+			ctx := context.Background()
+
+			// test if the attestation of the **taskrun** artifact can be successfully stored into grafeas server
+			// and test if payloads and signatures inside the attestation can be retrieved.
+			testInterface(ctx, t, test, backend, test.args.trPayload, test.args.trSignature, test.args.trOpts)
+
+			// test if the attestation of the **oci** artifact can be successfully stored into grafeas server
+			// and test if payloads and signatures inside the attestation can be retrieved.
+			testInterface(ctx, t, test, backend, test.args.ociPayload, test.args.ociSignature, test.args.ociOpts)
+
+			// test if all occurrences generated from `StorePayload` are what we expect.
+			testListOccurrences(ctx, t, backend)
+		})
+	}
+
+	// close connection
+	conn.Close()
+}
+
+// test attestation storage and retrieval
+func testInterface(ctx context.Context, t *testing.T, test testConfig, backend Backend, payload []byte, signature string, opts config.StorageOpts) {
+	if err := backend.StorePayload(ctx, payload, signature, opts); (err != nil) != test.wantErr {
+		t.Errorf("Backend.StorePayload() error = %v, wantErr %v", err, test.wantErr)
+	}
+
+	// get uri
+	objectIdentifier := backend.retrieveResourceURI(opts)
+
+	// check signature
+	expect_signature := map[string][]string{objectIdentifier: []string{signature}}
+	got_signature, err := backend.RetrieveSignatures(ctx, opts)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if !cmp.Equal(got_signature, expect_signature) {
+		t.Errorf("Wrong signature object received, got=%s", cmp.Diff(got_signature, expect_signature))
+	}
+
+	// check payload
+	expect_payload := map[string]string{objectIdentifier: string(payload)}
+	var got_payload map[string]string
+	got_payload, err = backend.RetrievePayloads(ctx, opts)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if !cmp.Equal(got_payload, expect_payload) {
+		t.Errorf("Wrong payload object received, got=%s", cmp.Diff(got_payload, expect_payload))
+	}
+}
+
+// test occurrences are generated correctly
+func testListOccurrences(ctx context.Context, t *testing.T, b Backend) {
+	wanted := &pb.ListOccurrencesResponse{
+		Occurrences: []*pb.Occurrence{
+			// occurrence for taskrun
+			{
+				// Occurrence Name will be automatically generated by grafeas server.
+				// In this fake grafeas server, we mock this behaviour by just using resource uri.
+				// In the real world, the occurrence name will be in the format of `projects/<PROJECT_NAME>/occurrences/<AUTO-GENERATED-ID>`.
+				// projects/my_project/occurrences/06d6e0d6-ee2b-4629-b44a-2188ac92eee4
+				Name: "/apis//namespaces/foo1//bar1@uid1",
+				// It's expected that the faked taskrun test does not have group version kind information.
```

Comment:
Shouldn't addTypeInformationToObject be filling this in?

Reply:
solved! a0c2680ac800e9c02c61ee5e1c287321e3e93363